src/online2/online-nnet2-feature-pipeline.cc

```diff
@@ -89,16 +122,39 @@ OnlineNnet2FeaturePipeline::OnlineNnet2FeaturePipeline(
     KALDI_ERR << "Code error: invalid feature type " << info_.feature_type;
   }
 
+  // Apply CMVN to features
+  if (info_.use_cmvn) {
+    KALDI_ASSERT(global_cmvn_stats_.NumRows() != 0);
+    if (info_.add_pitch) {
```

Comment:
Why do you handle 'pitch' differently? In the training scripts, the online-cmn is used on the 'base' vector augmented with pith. Briefly, the cmn is done also for pitch features. 

For the ivector extractor, this needs to get the unnormalized features, the online-cmvn is done inside the class 'OnlineIvectorFeature' by re-routing the feature stream that is going to the UBM-GMM model.

So you need to normalize the inputs of 'nnet3' model and that's it...
K.

Reply:
I had not noticed that they were not applying the CMN to the pitch.
They are using some model that they got from someone else, I think, to test their setup and their training scripts were not checked in.  They may have been doing the CMN in a different way to how we are doing it.
Incidentally, it's possible to turn off the normalization of the pitch features by supplying the appropriate --skip-dims option in the CMVN config.  So it should probably be safe to move this to after the pitch is added.  With suitable modifications to the config files, and possibly adding 3 dummy columns to the global-cmvn-stats file, it should be possible to get the behavior they want with that model.

Reply:
Normalize after pitch, please.  In online2-online-feature-pipeline.cc, I think the reason it's done that way is that the cmvn is only used get the posteriors for the i-vector extractor, which doesn't use the pitch at all.
So: apply before the pitch, please.

BTW, I think I mentioned previously that we'd want an option in the ivector stuff to apply the CMVN also to the features used to get the ivector stats (not just the posteriors), in the case where we are using CMN on the nnet-inptu features.    Karel used the option `online_cmvn_iextractor` in his scripts, and he had a boolean option called online_cmvn_iextractor in his code.  You would probably want to merge with his branch.  Can you create a version of this branch that includes Karel's changes?

@freewym can you please take note of this conversation?  Since you were going to help test this?

Reply:
Ok. BTW, I have merged these 2 PRs into my wake-word branch, and the the experiment is running.